### PR TITLE
fetchpatch: add throwBool, reason, condition, path

### PR DESCRIPTION
**Pull request by Artturin**
NOTE: i will write tests and docs later for now i want feedback about the namings
to view the diff well, click the `hide whitespace changes` button

`throwBool = lib.versionAtLeast version "2.1.0";`
so people won't forget to remove conditional patches

`condition = stdenv.isAarch64;`
so no `lib.optional` needed so we can evaluate the throwBool

reason = "${pname}: upstream fix for ncurses-6.3";
so we dont have to use comments

`path = ./some.patch;`
to allow using the above options with vendored patches

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/fetchpatch/default.nix [hunk 1]
```diff
@@ -15,6 +15,10 @@ in
 , excludes ? []
 , includes ? []
 , revert ? false
+, throwBool ? false
```
Comment (pennae) [inline pkgs/build-support/fetchpatch/default.nix:18]: not sure that communicates well what this argument does. would go for something like `validIf ? true` to make that part of a fetchpatch call easier to understand, but that doesn't really explain the "throw unless validIf" either. `throwUnless` maybe? not sure.

**Comment by ncfavier** [comment]
> `throwBool = lib.versionAtLeast version "2.1.0";` so people won't forget to remove conditional patches

Won't already applied patches cause the build to fail? I think those build errors are clear enough: "Reversed (or previously applied) patch detected!"

> `condition = stdenv.isAarch64;` so no `lib.optional` needed so we can evaluate the throwBool

I don't understand this, can you elaborate?

File: pkgs/build-support/fetchpatch/default.nix [hunk 2]
```diff
@@ -27,57 +31,63 @@ let
...
-in
-lib.throwIfNot (excludes == [] || includes == [])
-  "fetchpatch: cannot use excludes and includes simultaneously"
-fetchurl ({
-  postFetch = ''
-    tmpfile="$TMPDIR/patch"
+in let
+  fetch = if !condition then ""
+  else if builtins.isPath path then path
+  else fetchurl ({
```
Comment (roberth) [inline pkgs/build-support/fetchpatch/default.nix:37]: So it's not a fetcher anymore. It's a more general function that should have a general name.
When it's a patch from file, you could still return a package-like object that contains the added metadata.

File: pkgs/build-support/fetchpatch/default.nix [hunk 3]
```diff
@@ -27,57 +31,63 @@ let
...
+        echo "Normalized patch file was:" 1>&2
+        cat "$tmpfile" 1>&2
+        exit 1
+      fi
+    '' + lib.optionalString revert ''
+      ${patchutils}/bin/interdiff "$out" /dev/null > "$tmpfile"
+      mv "$tmpfile" "$out"
+    '' + postFetch;
+  } // builtins.removeAttrs args [
+    "relative" "stripLen" "extraPrefix" "excludes" "includes" "revert" "postFetch" "reason" "throwBool" "condition" "path"
```
Comment (roberth) [inline pkgs/build-support/fetchpatch/default.nix:86]: We shouldn't be programming like this. (Not your fault, I know)

File: pkgs/games/nudoku/default.nix [hunk 4]
```diff
@@ -12,11 +12,12 @@ stdenv.mkDerivation rec {
   };
 
   patches = [
-    # Pull upstream fix for ncurses-6.3
     (fetchpatch {
       name = "ncurses-6.3.patch";
       url = "https://github.com/jubalh/nudoku/commit/93899a0fd72e04b9f257e5f54af53466106b5959.patch";
       sha256 = "1h3za0dnx8fk3vshql5mhcici8aw8j0vr7ra81p3r1rii4c479lm";
+      reason = "${pname}: upstream fix for ncurses-6.3";
```
Comment (roberth) [inline pkgs/games/nudoku/default.nix:19]: `${pname}` should be redundant. Can we use `builtins.addErrorContext` in `mkDerivation`, or isn't the trace already good enough?

**Review by roberth — COMMENTED** [review]
It's not 100% clear to me what problem this all solves.

**Comment by Artturin** [comment]
thanks for the feedback, i will try to figure out a better way to add metadata to a patch

File: pkgs/build-support/fetchpatch/default.nix (hunk 3 above)
Comment (Artturin) [reply pkgs/build-support/fetchpatch/default.nix:86]: do you have an ideal example in nixpkgs?

Comment (roberth) [reply pkgs/build-support/fetchpatch/default.nix:86]: Everything that's already separate.
Example: packages and package sets. If I create a new package `foo`, I don't have to add `remoteAttrs args ["foo"]` to `callPackage`.
Something that's not been done yet: a `mkDerivation` that doesn't try to forward _every_ argument to the builder as an environment variable / drvAttr. I've seen the structured attrs PR contain what seems to be something of a fix for this, but that seems to be too good to be true.

**Pull request closed without merging**